File: examples/vec_store.c
/*
 * Variable-length "vec" records packed into a bstack.
 *
 * Layout per record:
 *   [length: uint32_t LE][data: length bytes][0x00 padding]
 *
 * The sentinel lets a reader skip past unknown data and find the next record
 * boundary without a separate index. The same framing pattern generalises to
 * variable-size nodes in graph or tree structures: prefix each node with its
 * byte length, append a 0x00 sentinel, and a linear scan walks every node
 * without needing an external index or fixed-size slots.
 */

#include "../c/bstack.h"

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

typedef bstack_t vec_store_t;

/* Decode a little-endian uint32 from four bytes. */
static uint32_t le32(const uint8_t b[4])
{
    return (uint32_t)b[0]
         | ((uint32_t)b[1] << 8)
         | ((uint32_t)b[2] << 16)
         | ((uint32_t)b[3] << 24);
}

/* Append a record. Returns its logical start offset, or (uint64_t)-1 on error. */
static uint64_t vs_push(vec_store_t *vs, const uint8_t *data, size_t len)
{
    uint8_t *record = malloc(5 + len);
    if (!record) return (uint64_t)-1;
    uint32_t ulen = (uint32_t)len;
    record[0] = (uint8_t)(ulen);
    record[1] = (uint8_t)(ulen >> 8);
    record[2] = (uint8_t)(ulen >> 16);
    record[3] = (uint8_t)(ulen >> 24);
    memcpy(record + 4, data, len);
    record[4 + len] = 0x00;
    uint64_t offset;
    int r = bstack_push(vs, record, 5 + len, &offset);
    free(record);
    return r == 0 ? offset : (uint64_t)-1;
}

/*
 * Read the record starting exactly at pos.
 * Fills buf with up to buf_size bytes of data; sets *data_len to the
 * record's actual data length. Returns the start offset of the next
 * record, or (uint64_t)-1 on error.
 */
static uint64_t vs_read_at(vec_store_t *vs, uint64_t pos,
                            uint8_t *buf, size_t buf_size, size_t *data_len)
{
    uint8_t lb[4];
    if (bstack_get(vs, pos, pos + 4, lb) != 0) return (uint64_t)-1;
    uint32_t len = le32(lb);
    *data_len = len;
    if (len > 0) {
        size_t copy = len < buf_size ? len : buf_size;
        if (bstack_get(vs, pos + 4, pos + 4 + copy, buf) != 0)
            return (uint64_t)-1;
    }
    return pos + 4 + (uint64_t)len + 1;
}

/*
 * Walk to the nth record (0-indexed) and fill buf with its data.
 * Returns 0 on success, -1 on error.
 */
static int vs_get_nth(vec_store_t *vs, size_t n,
                      uint8_t *buf, size_t buf_size, size_t *data_len)
{
    uint64_t pos = 0;
    for (size_t i = 0; i < n; i++) {
        uint8_t lb[4];
        if (bstack_get(vs, pos, pos + 4, lb) != 0) return -1;
        pos += 4 + (uint64_t)le32(lb) + 1;
    }
    uint64_t next = vs_read_at(vs, pos, buf, buf_size, data_len);
    return next != (uint64_t)-1 ? 0 : -1;
}

/*
 * Scan from the beginning and return the first record whose start offset
 * is >= min_pos. Sets *found_at and fills buf. Returns 0 on success, -1
 * if no such record exists.
 */
static int vs_get_after(vec_store_t *vs, uint64_t min_pos,
                        uint64_t *found_at,
                        uint8_t *buf, size_t buf_size, size_t *data_len)
{
    uint64_t total;
    if (bstack_len(vs, &total) != 0) return -1;
    uint64_t cur = 0;
    while (cur < total) {
        uint8_t lb[4];
        if (bstack_get(vs, cur, cur + 4, lb) != 0) return -1;
        uint32_t len = le32(lb);
        if (cur >= min_pos) {
            *found_at = cur;
            *data_len = len;
            if (len > 0) {
                size_t copy = len < buf_size ? len : buf_size;
                if (bstack_get(vs, cur + 4, cur + 4 + copy, buf) != 0)
                    return -1;
            }
            return 0;
        }
        cur += 4 + (uint64_t)len + 1;
    }
    return -1;
}

int main(void)
{
    vec_store_t *store = bstack_open("vec_store_example.bstack");
    if (!store) { perror("bstack_open"); return 1; }

    const char *records[] = { "alpha", "bb", "ccc", "dddd" };
    uint64_t offsets[4];
    for (int i = 0; i < 4; i++) {
        offsets[i] = vs_push(store,
                             (const uint8_t *)records[i],
                             strlen(records[i]));
    }
    printf("record start offsets: [%llu, %llu, %llu, %llu]\n",
           (unsigned long long)offsets[0], (unsigned long long)offsets[1],
           (unsigned long long)offsets[2], (unsigned long long)offsets[3]);

    uint64_t total;
    bstack_len(store, &total);
    printf("total store size:     %llu bytes\n", (unsigned long long)total);

    /* Read by index */
    printf("\nby index:\n");
    uint8_t buf[64];
    for (int i = 0; i < 4; i++) {
        size_t data_len;
        vs_get_nth(store, (size_t)i, buf, sizeof buf, &data_len);
        printf("  [%d] \"%.*s\"\n", i, (int)data_len, buf);
    }

    /* Sequential scan using vs_read_at */
    printf("\nsequential scan:\n");
    uint64_t pos = 0;
    while (pos < total) {
        size_t data_len;
        uint64_t next = vs_read_at(store, pos, buf, sizeof buf, &data_len);
        printf("  @ %4llu: \"%.*s\"\n",
               (unsigned long long)pos, (int)data_len, buf);
        pos = next;
    }

    /* Find first record at or after a raw file offset */
    uint64_t search_from = offsets[1] + 1;
    uint64_t found_at;
    size_t data_len;
    vs_get_after(store, search_from, &found_at, buf, sizeof buf, &data_len);
    printf("\nfirst record at or after offset %llu: \"%.*s\" (starts at %llu)\n",
           (unsigned long long)search_from,
           (int)data_len, buf,
           (unsigned long long)found_at);

    bstack_close(store);
    return 0;
}

File: examples/vec_store.rs
//! Variable-length "vec" records packed into a bstack.
//!
//! ## Layout
//!
//! ```text
//! ┌──────────┬─────────────────┬─────┬──────────┬────────┬─────┐
//! │ len: u32 │ data: len bytes │ 0x00│ len: u32 │  data  │ 0x00│ ...
//! └──────────┴─────────────────┴─────┴──────────┴────────┴─────┘
//! ^-- record 0 starts here          ^-- record 1 starts here
//! ```
//!
//! Each record is a little-endian u32 length, followed by that many data bytes,
//! followed by **at least one** 0x00 padding byte. The sentinel lets a reader
//! skip past unknown data and find the next record boundary without a separate
//! index.
//!
//! The same framing pattern generalises to variable-size nodes in graph or tree
//! structures stored in a flat file: prefix each node with its byte length,
//! append a 0x00 sentinel, and a linear scan can walk every node without
//! needing an external index or fixed-size slots. This gives you a rudimentary
//! variable-node-size persistent data structure on top of any append-only log.

use bstack::BStack;
use std::io;
use std::path::Path;

struct VecStore(BStack);

impl VecStore {
    fn open(path: impl AsRef<Path>) -> io::Result<Self> {
        Ok(VecStore(BStack::open(path)?))
    }

    /// Append a record. Returns its logical start offset in the store.
    fn push_vec(&self, data: &[u8]) -> io::Result<u64> {
        let len = data.len() as u32;
        let mut record = Vec::with_capacity(5 + data.len());
        record.extend_from_slice(&len.to_le_bytes());
        record.extend_from_slice(data);
        record.push(0x00); // padding sentinel
        self.0.push(&record)
    }

    /// Read the record whose header starts exactly at `pos`.
    /// Returns `(data, next_record_pos)`.
    fn read_vec_at(&self, pos: u64) -> io::Result<(Vec<u8>, u64)> {
        let mut len_buf = [0u8; 4];
        self.0.get_into(pos, &mut len_buf)?;
        let len = u32::from_le_bytes(len_buf) as u64;
        let data = self.0.get(pos + 4, pos + 4 + len)?;
        Ok((data, pos + 4 + len + 1))
    }

    /// Walk to the nth record (0-indexed) and return its data.
    fn get_nth_vec(&self, n: usize) -> io::Result<Vec<u8>> {
        let mut pos = 0u64;
        for _ in 0..n {
            let mut len_buf = [0u8; 4];
            self.0.get_into(pos, &mut len_buf)?;
            let len = u32::from_le_bytes(len_buf) as u64;
            pos += 4 + len + 1;
        }
        let (data, _) = self.read_vec_at(pos)?;
        Ok(data)
    }

    /// Scan from the beginning and return the first record whose start offset
    /// is >= `pos`, along with its offset.
    fn get_vec_after(&self, pos: u64) -> io::Result<(u64, Vec<u8>)> {
        let total = self.0.len()?;
        let mut cur = 0u64;
        while cur < total {
            let mut len_buf = [0u8; 4];
            self.0.get_into(cur, &mut len_buf)?;
            let len = u32::from_le_bytes(len_buf) as u64;
            if cur >= pos {
                let data = self.0.get(cur + 4, cur + 4 + len)?;
                return Ok((cur, data));
            }
            cur += 4 + len + 1;
        }
        Err(io::Error::new(
            io::ErrorKind::NotFound,
            "no record at or after given position",
        ))
    }
}

fn main() -> io::Result<()> {
    let store = VecStore::open("vec_store_example.bstack")?;

    // Push variable-length records.
    let offsets = [
        store.push_vec(b"alpha")?,
        store.push_vec(b"bb")?,
        store.push_vec(b"ccc")?,
        store.push_vec(b"dddd")?,
    ];
    println!("record start offsets: {:?}", offsets);
    println!("total store size:     {} bytes", store.0.len()?);

    // Read by index.
    println!("\nby index:");
    for i in 0..4 {
        let data = store.get_nth_vec(i)?;
        println!("  [{}] {:?}", i, String::from_utf8_lossy(&data));
    }

    // Sequential scan using read_vec_at.
    println!("\nsequential scan:");
    let mut pos = 0u64;
    let total = store.0.len()?;
    while pos < total {
        let (data, next) = store.read_vec_at(pos)?;
        println!("  @ {:>4}: {:?}", pos, String::from_utf8_lossy(&data));
        pos = next;
    }

    // Find first record at or after a raw file offset.
    let search_from = offsets[1] + 1; // somewhere inside record 1
    let (found_at, data) = store.get_vec_after(search_from)?;
    println!(
        "\nfirst record at or after offset {}: {:?} (starts at {})",
        search_from,
        String::from_utf8_lossy(&data),
        found_at,
    );

    Ok(())
}
